app/models/taxon_scoring_model.rb

```diff
@@ -0,0 +1,96 @@
+# Scoring Model for Taxon
+# the scoring model is represented in json with a tree structure where a parent will apply an operator to all the children and the leaf node could be a constant(const) or an attribute from the taxon (attr)
+# Currently the operators supported are addition (+), product (-) for nodes with multiple children and atomic transformation function for nodes with one children.
+# Here is an example of how the following scoring function can be translated to a json
+#
+# aggregate_score = |genus.NT.zscore|*species.NT.zscore*species.NT.rpm + |genus.NR.zscore|*species.NR.zscore*species.NR.rpm
+# {
+#  "op" : "+",
+#  "on" : [
+#    {
+#      "op": "*",
+#      "on": [
+#        {
+#          "op": "abs",
+#          "on": { "op": "attr", "on": "genus.NT.zscore" }
+#        },
+#        { "op": "attr", "on": "species.NT.zscore" },
+#        { "op": "attr", "on": "species.NT.rpm" }
+#      ]
+#    },
+#    {
+#      "op": "*",
+#      "on": [
+#        {
+#          "op": "abs",
+#          "on": { "op": "attr", "on": "genus.NR.zscore" }
+#        },
+#        { "op": "attr", "on": "species.NR.zscore" },
+#        { "op": "attr", "on": "species.NR.rpm" }
+#      ]
+#    }
+#  ]
+# }
+class TaxonScoringModel < ApplicationRecord
+  include Math
+  attr_accessor :model
+  before_save :set_json
+  after_find  :set_model
+
+  DEFAULT_MODEL_NAME = 'aggregate_score'.freeze
+
+  def set_model
+    self.model = JSON.parse(model_json)
+  end
+
+  def set_json
+    self.model_json = model.to_json
+  end
+
+  def model_json=(json)
+    super(json)
+    self.model = JSON.parse(json)
+  end
+
+  def self.flatten_taxon(taxon_info, prefix = "")
+    result = {}
+    taxon_info.each do |key, val|
+      if val.is_a?(Hash)
+        new_prefix = "#{prefix}#{key}."
+        result.merge!(flatten_taxon(val, new_prefix))
+      else
+        result["#{prefix}#{key}"] = val
+      end
+    end
+    result
+  end
+
+  def score(taxon, score_func = model)
+    op = score_func["op"] # return 0.0 might not be the best solution but could avoid fatal error TODO(yf): revisit
+    return 0.0 unless op
+    case op
+    when "+"
+      result = 0.0
+      score_func["on"].each { |sf| result += score(taxon, sf) }
+      return result
+    when "*"
+      result = 1.0
+      score_func["on"].each { |sf| result *= score(taxon, sf) }
+      return result
+    when "const"
+      return score_func["on"].to_f
+    when "attr"
+      return taxon[score_func["on"]].to_f
+    else
+      # atomic transformation function like abs, sqrt, log10
+      result = score(taxon, score_func["on"])
+      begin
+        # eg. result.abs
+        return result.send(op)
```

Comment:
Could you validate that the ops are safe before you do send?   I actually thought this ability to validate, for security purposes, was the chief advantage of interpreting the json AST formula yourself, rather than just getting a big math formula and sending it all to ruby.  If this is addressed, I think I'd be happy to approve the pull request.

Reply:
OK. I have issues with this too. I will try coming up a solution.